### PR TITLE
Struct sensor_id

### PR DESCRIPTION
**Pull request by gaschler**
Assign RFC=0012 to struct sensor_id.

### CODE REVIEW

File: text/0000-struct-sensor-id.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+When distributing local and global slam over different processes, identifying sensors by a string is no longer sufficient, because multiple range finder sensors in local SLAM become a single LocalSlamResult sensor on the global side.
+For example, when local SLAM runs on an agent server and global SLAM on an uplink server, the agent server expects n range finder sensors and additional sensors, while the collator of the uplink server waits for only one "LocalSlamResultData" input and the addtional sensors.
+
+## Approach
+[approach]: #approach
+
+The suggested change is to replace 'std::string sensor_id' by a struct that contains an 'enum SensorType' and a 'std::string id'.
+This allows us to implement distributed local/global mapping in a clean way, such as removing range finder sensors from the set and adding a sensor type for local slam results.
+
+As an additional benefit, future concepts about properties of sensors would be easy to implement.
```
Comment (cschuet) [inline text/0000-struct-sensor-id.md]: Can you give some examples, e.g. tagging sparse / unreliable sensors.

File: text/0000-struct-sensor-id.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# Struct sensor_id
+
+## Summary
+[summary]: #summary
+
+Replace the string sensor_id by a structure with an enum sensor type and a string.
```
Comment (cschuet) [inline text/0000-struct-sensor-id.md]: The way I think about this is to add the possibility to attach sensor metadata to a the sensor_id.

**Review by cschuet — APPROVED** [review]

**Comment by ojura** [comment]
How is cartographer_ros going to follow this?

Please have a look at `sensor_topics` in [RFC 0009](https://github.com/googlecartographer/rfcs/blob/master/text/0009-offline-node-improved-trajectories-bags.md#data-format-of-the-configuration), which is purely string-based.

I am waiting for a review of googlecartographer/cartographer_ros#636 which implements RFC 0009. I would appreciate if #636 got merged before googlecartographer/cartographer#839, so I don't have to make further updates to #636.

**Comment by gaschler** [comment]
@ojura I agree that https://github.com/googlecartographer/cartographer_ros/pull/636 should go first.

For this change, cartographer_ros will just add the new SensorType information depending on the 
Ros MessageType, so this should be straightforward. However, I haven't actually written the cartographer_ros follow-up yet.

**Comment by ojura** [comment]
Ok. You don’t know the message type though until you receive one; isn’t the trajectory already supposed to be started before that point?

**Comment by gaschler** [comment]
@ojura I am planning to use `num_laser_scans = 0` and so on from the lua config of cartographer_ros.

File: text/0000-struct-sensor-id.md (hunk 1 above)
Comment (gaschler) [reply text/0000-struct-sensor-id.md]: Done

File: text/0000-struct-sensor-id.md (hunk 2 above)
Comment (gaschler) [reply text/0000-struct-sensor-id.md]: Done

File: text/0000-struct-sensor-id.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+## Motivation
+[motivation]: #motivation
+
+When distributing local and global slam over different processes, identifying sensors by a string is no longer sufficient, because multiple range finder sensors in local SLAM become a single LocalSlamResult sensor on the global side.
+For example, when local SLAM runs on an agent server and global SLAM on an uplink server, the agent server expects n range finder sensors and additional sensors, while the collator of the uplink server waits for only one "LocalSlamResultData" input and the addtional sensors.
+
+## Approach
+[approach]: #approach
+
+The suggested change is to replace 'std::string sensor_id' by a struct that contains an 'enum SensorType' and a 'std::string id'.
```
Comment (SirVer) [inline text/0000-struct-sensor-id.md]: backticks for code?

**Review by SirVer — APPROVED** [review]
makes sense to me.

File: text/0000-struct-sensor-id.md (hunk 3 above)
Comment (gaschler) [reply text/0000-struct-sensor-id.md]: Done

File: text/0000-struct-sensor-id.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+## Summary
+[summary]: #summary
+
+Replace the `string` sensor_id by a `struct` which allows us to include sensor metadata, such as an `enum SensorType`.
+
+## Motivation
+[motivation]: #motivation
+
+When distributing local and global slam over different processes, identifying sensors by a string is no longer sufficient, because multiple range finder sensors in local SLAM become a single LocalSlamResult sensor on the global side.
+For example, when local SLAM runs on an agent server and global SLAM on an uplink server, the agent server expects n range finder sensors and additional sensors, while the collator of the uplink server waits for only one "LocalSlamResultData" input and the addtional sensors.
```
Comment (pifon2a) [inline text/0000-struct-sensor-id.md]: nit: s/expects n/expects `n`/

**Review by pifon2a — APPROVED** [review]

File: text/0000-struct-sensor-id.md (hunk 4 above)
Comment (gaschler) [reply text/0000-struct-sensor-id.md]: Done

**Pull request merged**